staging/src/k8s.io/apiserver/pkg/audit/request.go

```diff
@@ -244,3 +284,62 @@ func maybeTruncateUserAgent(req *http.Request) string {
 
 	return ua
 }
+
+// copyWithoutManagedFields will make a deep copy of the specified object and will discard
```

Comment:
why does the copy have to be deep? I'd expect shallow is fine for everything not a direct ancestor of managed fields?

Reply:
I am not certain if shallow copy is safe and has no side effect. If you think it's safe I can remove the deep copy.

Reply:
I can't think of a reason why a shallow copy would be problematic here. @wojtek-t for a second opinion, but I think avoiding a deep copy on every object that's audited would be really good.

Reply:
do we have a way to do a shallow copy of an interface object reliably?

Reply:
> I am not certain if shallow copy is safe and has no side effect

I spoke too soon, a shallow copy should be safe since we are dropping `managedFields` slice. I did not come across any utility in k8s to shallow copy an interface. There might be a way to use `reflect` (will introduce some cpu overhead) to build a shallow copy, I will look into it.

Reply:
I did not come across any utility in k8s to shallow copy from an interface object. One option is to use the `reflect` package, traverse all the fields of the type recursively and build a shallow copy. I am not sue if this is the route we want to go.

I don't know if the serializer can be tweaked to omit certain fields or one can provide custom `MarshalJSON` function. But such extension is not worth pursuing in my opinion, since it can lead to bugs in the system due to improper use.

@lavalamp @sttts @liggitt please let me know how you want me to move forward with this.

Reply:
If it's excessively hard it's probably worth just doing the easier thing and measuring to see if it's actually a performance problem.

Reply:
1. I agree that deep-copy isn't really needed here - shallow is enough.

2. I can live with deep-copy for now, but please add a TODO to figure out how to migrate to shallow copy here.
I don't have an easy way from the top of my head to that in a generic case.